{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f4e78dde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Boiler plate stuff to start the module\n",
    "import jpype\n",
    "import jpype.imports\n",
    "from jpype.types import *\n",
    "import sys \n",
    "import numpy as np\n",
    "import traceback"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5077507e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Launch the JVM\n",
    "jpype.startJVM(classpath=['./maze-server-v2019.4.jar'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "22f4292f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the Java modules\n",
    "from java.io import IOException\n",
    "from java.net import Socket\n",
    "from java.net import UnknownHostException\n",
    "\n",
    "from javax.net.ssl import SSLSocket\n",
    "from javax.net.ssl import SSLSocketFactory\n",
    "from javax.xml.bind import JAXBException\n",
    "from javax.xml.bind import UnmarshalException\n",
    "\n",
    "from de.fhac.mazenet.server.generated import AwaitMoveMessageData\n",
    "from de.fhac.mazenet.server.generated import Errortype\n",
    "from de.fhac.mazenet.server.generated import MazeCom\n",
    "from de.fhac.mazenet.server.generated import MazeComMessagetype\n",
    "from de.fhac.mazenet.server.generated import MoveMessageData\n",
    "from de.fhac.mazenet.server.generated import BoardData\n",
    "\n",
    "from de.fhac.mazenet.server.networking import MazeComMessageFactory\n",
    "from de.fhac.mazenet.server.networking import XmlInputStream\n",
    "from de.fhac.mazenet.server.networking import XmlOutputStream\n",
    "from de.fhac.mazenet.server.game import Board\n",
    "from de.fhac.mazenet.server.game import Position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bde3ecff",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Client:\n",
    "    __id = None\n",
    "    def __init__(self, name, address, port, tls):\n",
    "        if(tls):\n",
    "            self.sslSocketFactory = SSLSocketFactory.getDefault()\n",
    "            self.sslSocket = self.sslSocketFactory.createSocket(address, port)\n",
    "            self.in_ = XmlInputStream(self.sslSocket.getInputStream())\n",
    "            self.out = XmlOutputStream(self.sslSocket.getOutputStream())\n",
    "        else:\n",
    "            self.socket = Socket(address, port)\n",
    "            self.in_ = XmlInputStream(self.socket.getInputStream())\n",
    "            self.out = XmlOutputStream(self.socket.getOutputStream())\n",
    "\n",
    "        self.name = name\n",
    "        self.instance = None\n",
    "    \n",
    "    def getId(self):\n",
    "        return self.__id\n",
    "    \n",
    "    def resetId(self):\n",
    "        self.__id = 0\n",
    "\n",
    "    @staticmethod\n",
    "    def getInstance(self, name, address, port, tls):\n",
    "        if(self.instance == None): self.instance = Client(name, address, port, tls)\n",
    "        return self.instance\n",
    "\n",
    "    def login(self):\n",
    "        login = MazeComMessageFactory.createLoginMessage(self.name)\n",
    "        self.out.write(login)\n",
    "        loginResponse = self.in_.readMazeCom()\n",
    "        if(loginResponse.getMessagetype() == MazeComMessagetype.LOGINREPLY):\n",
    "            self.__id = loginResponse.getLoginReplyMessage().getNewID()\n",
    "        elif(loginResponse.getMessagetype() == MazeComMessagetype.ACCEPT):\n",
    "            print(\"Login Fehler: Falsche Nachricht\")\n",
    "            sys.exit(1)\n",
    "        elif (loginResponse.getMessagetype() == MazeComMessagetype.DISCONNECT):\n",
    "            print(\"Login Fehler: Loginversuche\")\n",
    "            sys.exit(1)\n",
    "        else:\n",
    "            print(\"Login Fehler: bekommte Message Type\" + loginResponse.getMessagetype())\n",
    "            sys.exit(1)\n",
    "\n",
    "    def play(self):\n",
    "        while(True):\n",
    "            try:\n",
    "                receivedMazeCom = self.in_.readMazeCom()\n",
    "                if(receivedMazeCom.getMessagetype() == MazeComMessagetype.AWAITMOVE):\n",
    "                    self.awaitMove(receivedMazeCom)\n",
    "                elif(receivedMazeCom.getMessagetype() == MazeComMessagetype.ACCEPT):\n",
    "                    self.accept(receivedMazeCom.getAcceptMessage().getErrortypeCode())\n",
    "                elif(receivedMazeCom.getMessagetype() == MazeComMessagetype.DISCONNECT):\n",
    "                    print(\"You have been disconnected\")\n",
    "                    self.disconnect(receivedMazeCom.getDisconnectMessage().getErrortypeCode())\n",
    "                elif(receivedMazeCom.getMessagetype() == MazeComMessagetype.MOVEINFO):\n",
    "                    print(\"in MoveInfo\")\n",
    "                elif(receivedMazeCom.getMessagetype() == MazeComMessagetype.WIN):\n",
    "                    print(\"You have won\")\n",
    "                    sys.exit(0)\n",
    "                else:\n",
    "                    print(\"Unkown message type: \" + receivedMazeCom.getMessagetype())\n",
    "            except Exception:\n",
    "                print(traceback.format_exc())\n",
    "\n",
    "    def accept(self, errortype):\n",
    "        if(errortype ==  Errortype.NOERROR):\n",
    "            print(\"Alles Ok, Bewegung erlaubt.\")\n",
    "        elif(errortype ==  Errortype.AWAIT_MOVE):\n",
    "            print(\"Falsche Nachricht, Bewegung nicht erlaubt.\")\n",
    "        elif(errortype ==  Errortype.ILLEGAL_MOVE):\n",
    "            print(\"Regelversto√ü, unerlaubte Bewegung.\")\n",
    "        else:\n",
    "            print(\"Unbekannte Fehlermeldung: \" + errortype)\n",
    "    \n",
    "    def disconnect(self, errortype):\n",
    "        if(errortype ==  Errortype.TOO_MANY_TRIES):\n",
    "            print(\"Zu viele Versuche. Geworfen vom Server.\")\n",
    "            sys.exit(1)\n",
    "        else:\n",
    "            print(\"Ende des Spiels.\")\n",
    "            sys.exit(0)\n",
    "\n",
    "    def awaitMove(self, receivedMazeCom):\n",
    "        awaitMove = receivedMazeCom.getAwaitMoveMessage()\n",
    "        allPossibleBoards = []\n",
    "        if(awaitMove != None):\n",
    "            boardData = awaitMove.getBoard()\n",
    "            board = Board(boardData)\n",
    "            potentialShiftMoves = []\n",
    "            for i in np.arange(1, 6, 2):\n",
    "                potentialShiftMoves.append(Position(0, i))\n",
    "                potentialShiftMoves.append(Position(6, i))\n",
    "                potentialShiftMoves.append(Position(i, 0))\n",
    "                potentialShiftMoves.append(Position(i, 6))\n",
    "            \n",
    "            potentialShiftMoves.remove(board.getForbidden())\n",
    "\n",
    "            move = MoveMessageData()\n",
    "            move.setShiftCard(boardData.getShiftCard())\n",
    "            move.setNewPinPos(board.findPlayer(self.__id))\n",
    "            move.setShiftPosition(potentialShiftMoves[0])\n",
    "            \n",
    "            mazeComToSend = MazeCom()\n",
    "            mazeComToSend.setId(self.__id)\n",
    "            mazeComToSend.setMessagetype(MazeComMessagetype.MOVE)\n",
    "            mazeComToSend.setMoveMessage(move)\n",
    "            self.out.write(mazeComToSend)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "afca3ff6",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOST = \"127.0.0.1\"\n",
    "PORT = 5123"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7fb73702",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gym \n",
    "from gym import Env\n",
    "from gym.spaces import Discrete, Box, Dict, Tuple, MultiBinary, MultiDiscrete \n",
    "import numpy as np\n",
    "import random\n",
    "import os\n",
    "from stable_baselines3 import PPO\n",
    "from stable_baselines3.common.vec_env import VecFrameStack\n",
    "from stable_baselines3.common.evaluation import evaluate_policy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a1c20ec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from mss import mss\n",
    "import pydirectinput\n",
    "import cv2\n",
    "import pytesseract\n",
    "import matplotlib.pyplot as plt\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3c1d1e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pytesseract.pytesseract.tesseract_cmd = r'C:\\Program Files\\Tesseract-OCR\\tesseract.exe'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ecf572f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "class MazeEnv(Env):\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "        # Actions we can take, rotate card left, rotate card right, push card, go up, down, left, right, no op\n",
    "        # actions: push card, move, no op\n",
    "        # self.action_space = Discrete(3)\n",
    "        # actions: pick x and y where the next card should be pushed\n",
    "        self.action_space = Box(low=np.array([0,0]), high=np.array([6,6]), dtype=np.int8)\n",
    "        #  array\n",
    "        #self.observation_space = Tuple((\n",
    "        #                             Discrete(50),\n",
    "        #                             Discrete(24),\n",
    "        #                             Discrete(2))\n",
    "        #                            )\n",
    "        self.observation_space = Box(low=0, high=255, shape=(1,800,1100), dtype=np.uint8)\n",
    "        self.cap = mss()\n",
    "        self.game_location = {'top':200, 'left':100, 'width':1700, 'height':800}\n",
    "        self.end_location = {'top':50, 'left':400, 'width':400, 'height':150}\n",
    "        self.client = 0\n",
    "    \"\"\"    \n",
    "        self.players = np.array([0,0,0,0])\n",
    "        \n",
    "    def add_player(self, client):\n",
    "        for i in range(4):\n",
    "            if self.players[i] == 0:\n",
    "                self.players[i] = client\n",
    "                break\n",
    "    \"\"\"    \n",
    "    def step(self, action):\n",
    "        reward = 0\n",
    "        #if self.client.getId() == 0:\n",
    "        #    self.client.login()\n",
    "        #    time.sleep(1000)\n",
    "        try:\n",
    "            receivedMazeCom = self.client.in_.readMazeCom()\n",
    "            if(receivedMazeCom.getMessagetype() == MazeComMessagetype.AWAITMOVE):\n",
    "                awaitMove = self.client.in_.readMazeCom().getAwaitMoveMessage()\n",
    "                if(awaitMove != None):\n",
    "                    boardData = awaitMove.getBoard()\n",
    "                    board = Board(boardData)\n",
    "\n",
    "                    move = MoveMessageData()\n",
    "                    move.setShiftCard(boardData.getShiftCard())\n",
    "                    move.setNewPinPos(board.findPlayer(client.getId()))\n",
    "                    move.setShiftPosition(Position(action[0], action[1]))\n",
    "\n",
    "                    mazeComToSend = MazeCom()\n",
    "                    mazeComToSend.setId(self.__id)\n",
    "                    mazeComToSend.setMessagetype(MazeComMessagetype.MOVE)\n",
    "                    mazeComToSend.setMoveMessage(move)\n",
    "                    self.client.out.write(mazeComToSend)\n",
    "\n",
    "                    reward += 10\n",
    "            elif(receivedMazeCom.getMessagetype() == MazeComMessagetype.ACCEPT):\n",
    "                self.client.accept(receivedMazeCom.getAcceptMessage().getErrortypeCode())\n",
    "                print(\"aaccepted\")\n",
    "                reward += 5\n",
    "            elif(receivedMazeCom.getMessagetype() == MazeComMessagetype.DISCONNECT):\n",
    "                print(\"You have been disconnected\")\n",
    "                reward -= 10\n",
    "                self.client.disconnect(receivedMazeCom.getDisconnectMessage().getErrortypeCode())\n",
    "            elif(receivedMazeCom.getMessagetype() == MazeComMessagetype.MOVEINFO):\n",
    "                print(\"in MoveInfo\")\n",
    "            elif(receivedMazeCom.getMessagetype() == MazeComMessagetype.WIN):\n",
    "                print(\"You have won\")\n",
    "                reward += 100\n",
    "                sys.exit(0)\n",
    "            else:\n",
    "                print(\"Unkown message type: \" + receivedMazeCom.getMessagetype())\n",
    "        except Exception:\n",
    "            print(traceback.format_exc())\n",
    "        #############\n",
    "        done, done_cap = self.get_done() \n",
    "        observation = self.get_observation()\n",
    "        info = {}\n",
    "        return observation, reward, done, info\n",
    "        \n",
    "    def render(self):\n",
    "        # Implement viz\n",
    "        plt.imshow(np.array(self.cap.grab(self.end_location)))\n",
    "    \n",
    "    def reset(self):\n",
    "        time.sleep(1)\n",
    "        # stop\n",
    "        pydirectinput.click(x=1550, y=200)\n",
    "        # start\n",
    "        pydirectinput.click(x=1450, y=200)\n",
    "        print(\"before sleep\")\n",
    "        time.sleep(10)\n",
    "        print(\"after sleeeeeeep\")\n",
    "        #self.client = Client(self.client.name, HOST, PORT, False)\n",
    "        self.client.login()\n",
    "        return self.get_observation()\n",
    "    \n",
    "    # Closes the game?\n",
    "    def close(self):\n",
    "        pass\n",
    "    \n",
    "    def get_observation(self):\n",
    "        raw = np.array(self.cap.grab(self.game_location))[:,:,:3]\n",
    "        \n",
    "        gray = cv2.cvtColor(raw, cv2.COLOR_BGR2GRAY) \n",
    "        \n",
    "        resized = cv2.resize(gray, (100,83))\n",
    "        channel = np.reshape(resized, (1,83,100))\n",
    "        return channel\n",
    "    \n",
    "    def get_done(self):\n",
    "        done_screen = np.array(self.cap.grab(self.end_location))\n",
    "        \n",
    "        done_strs = ['Winner', 'Winn']\n",
    "        \n",
    "        # ocr\n",
    "        done = False\n",
    "        psm = r'--psm 8'\n",
    "        res = pytesseract.image_to_string(done_screen, config=psm)[:4]\n",
    "        if res in done_strs:\n",
    "            done = True\n",
    "        \n",
    "        return done, done_screen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5536aa85",
   "metadata": {},
   "outputs": [],
   "source": [
    "menv = MazeEnv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6e67fb83",
   "metadata": {},
   "outputs": [],
   "source": [
    "menv.client = Client('joe', HOST, PORT, False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63b6e402",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before sleep\n",
      "after sleeeeeeep\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n",
      "after sleep in looooop\n",
      "Traceback (most recent call last):\n",
      "  File \"XmlInputStream.java\", line 68, in de.fhac.mazenet.server.networking.XmlInputStream.readMazeCom\n",
      "  File \"UTFInputStream.java\", line 20, in de.fhac.mazenet.server.networking.UTFInputStream.readUTF8\n",
      "  File \"UTFInputStream.java\", line 41, in de.fhac.mazenet.server.networking.UTFInputStream.readNBytes\n",
      "Exception: Java Exception\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\AStA\\AppData\\Local\\Temp\\ipykernel_12592\\444313929.py\", line 35, in step\n",
      "    receivedMazeCom = self.client.in_.readMazeCom()\n",
      "java.io.java.io.EOFException: java.io.EOFException: [Fehler]: Konnte nur 0 von 4 angefordeten Bytes lesen\n",
      "\n",
      "before sleep in looooop\n"
     ]
    }
   ],
   "source": [
    "obs = menv.reset()\n",
    "done = False  \n",
    "total_reward   = 0\n",
    "while not done: \n",
    "    obs, reward,  done, info =  menv.step(menv.action_space.sample())\n",
    "    print('before sleep in looooop')\n",
    "    time.sleep(10)\n",
    "    print('after sleep in looooop')\n",
    "    total_reward  += reward\n",
    "print('Total Reward for episode {} is {}'.format(episode, total_reward))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c56a6c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18af7d7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    c = Client('joe', HOST, PORT, False)\n",
    "    c.login()\n",
    "    c.play()\n",
    "except Exception:\n",
    "    print(traceback.format_exc())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7550c6e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# jpype.shutdownJVM()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
